### ISSUE TITLE
schema migration director_datalist_entry to 1.9 breaks 

### ISSUE THREAD

**Issue by flohoff** [Icinga/icingaweb2-module-director#2492]
Hi,
after migrating from a pretty old director version to 1.9 the director_datalist_entrys do not get extended with a UUID field. 
Using the datalist entries then break.

postgresql table running on the migrated icinga director 1.9

```
i2director=# \d director_datalist_entry
                         Table "public.director_datalist_entry"
    Column     |          Type          | Collation | Nullable |         Default         
---------------+------------------------+-----------+----------+-------------------------
 list_id       | integer                |           | not null | 
 entry_name    | character varying(255) |           | not null | 
 entry_value   | text                   |           |          | 
 format        | enum_property_format   |           |          | 
 allowed_roles | character varying(255) |           |          | NULL::character varying
Indexes:
    "director_datalist_entry_pkey" PRIMARY KEY, btree (list_id, entry_name)
    "datalist_entry_datalist" btree (list_id)
Foreign-key constraints:
    "director_datalist_entry_datalist" FOREIGN KEY (list_id) REFERENCES director_datalist(id) ON UPDATE CASCADE ON DELETE CASCADE
```

Backtrace in webfrontend:
```

SQLSTATE[42703]: Undefined column: 7 ERROR: column director_datalist_entry.uuid does not exist
LINE 1: SELECT director_datalist_entry.uuid FROM director_datalist_e...
^, query was: SELECT director_datalist_entry.uuid FROM director_datalist_entry WHERE ("list_id" = 6) AND ("entry_name" = 'ESX-RB-CL-PROD')

#0 /usr/share/icingaweb2/library/vendor/Zend/Db/Statement.php(297): Zend_Db_Statement_Pdo->_execute()
#1 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Abstract.php(470): Zend_Db_Statement->execute()
#2 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract->query()
#3 /usr/share/icingaweb2/library/vendor/Zend/Db/Adapter/Abstract.php(816): Zend_Db_Adapter_Pdo_Abstract->query()
#4 /usr/src/icingaweb2-module-director/library/Director/Db/Branch/UuidLookup.php(117): Zend_Db_Adapter_Abstract->fetchOne()
#5 /usr/src/icingaweb2-module-director/library/Director/Db/Branch/UuidLookup.php(91): Icinga\Module\Director\Db\Branch\UuidLookup::fetchOptionalUuid()
#6 /usr/src/icingaweb2-module-director/library/Director/Data/Db/DbObject.php(1377): Icinga\Module\Director\Db\Branch\UuidLookup::findUuidForKey()
#7 /usr/src/icingaweb2-module-director/library/Director/Web/Form/Validate/IsDataListEntry.php(53): Icinga\Module\Director\Data\Db\DbObject::exists()
#8 /usr/src/icingaweb2-module-director/library/Director/Web/Form/Validate/IsDataListEntry.php(39): Icinga\Module\Director\Web\Form\Validate\IsDataListEntry->isListEntry()
#9 /usr/share/icingaweb2/library/vendor/Zend/Form/Element.php(1428): Icinga\Module\Director\Web\Form\Validate\IsDataListEntry->isValid()
#10 /usr/share/icingaweb2/library/vendor/Zend/Form.php(2280): Zend_Form_Element->isValid()
#11 /usr/src/icingaweb2-module-director/library/Director/Web/Form/QuickForm.php(393): Zend_Form->isValid()
#12 /usr/src/icingaweb2-module-director/library/Director/Web/Controller/ObjectController.php(638): Icinga\Module\Director\Web\Form\QuickForm->handleRequest()
#13 /usr/src/icingaweb2-module-director/library/Director/Web/Controller/ObjectController.php(171): Icinga\Module\Director\Web\Controller\ObjectController->addObjectForm()
#14 /usr/src/icingaweb2-module-director/application/controllers/HostController.php(99): Icinga\Module\Director\Web\Controller\ObjectController->editAction()
#15 /usr/src/icingaweb2-module-director/library/Director/Web/Controller/ObjectController.php(117): Icinga\Module\Director\Controllers\HostController->editAction()
#16 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Director\Web\Controller\ObjectController->indexAction()
#17 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#18 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#19 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch()
#20 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch()
#21 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#22 {main}
```

**Comment by Yoda-BZH** [comment]
same with mysql :

```
MariaDB [icingadirector]> select MAX(schema_version) from director_schema_migration\G
*************************** 1. row ***************************
MAX(schema_version): 178
1 row in set (0.000 sec)

MariaDB [icingadirector]> show create table director_datalist_entry \G
*************************** 1. row ***************************
       Table: director_datalist_entry
Create Table: CREATE TABLE `director_datalist_entry` (
  `list_id` int(10) unsigned NOT NULL,
  `entry_name` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `entry_value` text DEFAULT NULL,
  `format` enum('string','expression','json') DEFAULT NULL,
  `allowed_roles` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`list_id`,`entry_name`),
  CONSTRAINT `director_datalist_value_datalist` FOREIGN KEY (`list_id`) REFERENCES `director_datalist` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8
1 row in set (0.000 sec)
```

**Comment by Yoda-BZH** [comment]
I "fixed" it by using:

```
ALTER TABLE director_datalist_entry ADD COLUMN uuid VARBINARY(16) DEFAULT NULL AFTER list_id;
SET @tmp_uuid = LOWER(CONCAT(
     LPAD(HEX(FLOOR(RAND()   * 0xffff)), 4, '0'),
     LPAD(HEX(FLOOR(RAND() * 0xffff)), 4, '0'), '-',
     LPAD(HEX(FLOOR(RAND() * 0xffff)), 4, '0'), '-',
     '4',
     LPAD(HEX(FLOOR(RAND() * 0x0fff)), 3, '0'), '-',
     HEX(FLOOR(RAND() * 4 + 8)),
     LPAD(HEX(FLOOR(RAND() * 0x0fff)), 3, '0'), '-',
     LPAD(HEX(FLOOR(RAND() * 0xffff)), 4, '0'),
     LPAD(HEX(FLOOR(RAND() * 0xffff)), 4, '0'),
     LPAD(HEX(FLOOR(RAND() * 0xffff)), 4, '0')
));
UPDATE director_datalist_entry SET uuid = UNHEX(LPAD(LPAD(HEX(CONCAT(list_id, MD5(entry_name))), 16, '0'), 32, REPLACE(@tmp_uuid, '-', ''))) WHERE uuid IS NULL;
ALTER TABLE director_datalist_entry MODIFY COLUMN uuid VARBINARY(16) NOT NULL, ADD UNIQUE INDEX uuid (uuid);
```

I used an uuid generation routine from another migration file. However, the `id` field is absent in this table. Using `CONCAT(list_id, entry_id)` gave me conflicting IDs because of string length.

I managed by using `MD5(entry_name)` to generate unique values with the limited string length.

**Comment by Yoda-BZH** [comment]
When trying to create a Backet snapshot of datalists, I have the MySQL error :
```
        Trying to set invalid director_datalist_entry key "uuid". DB schema change? (DbObject.php:731)
```

**Comment by Thomas-Gelf** [comment]
`director_datalist_entry` has no `uuid` column, at least not yet. Please undo your change, drop that handcrafted column and try again with v1.9.1. This issue should have been addressed with #2475

**Comment by Yoda-BZH** [comment]
:+1: 

Thank you !

**Comment by flohoff** [comment]
> `director_datalist_entry` has no `uuid` column, at least not yet. Please undo your change, drop that handcrafted column and try again with v1.9.1. This issue should have been addressed with #2475

So how comes the backtrace shows the code requiring it to exist? See the original backtrace i opened this issue with.

Flo
